workers/unity/Packages/io.improbable.worker.sdk/IOStream.cs

```diff
@@ -0,0 +1,120 @@
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using Improbable.Worker.CInterop.Internal;
+
+namespace Improbable.Worker.CInterop
+{
+    public unsafe class IOStream : IDisposable
+    {
+        private readonly CIO.StreamHandle stream;
+
+        private IOStream(CIO.StreamHandle stream)
+        {
+            this.stream = stream;
+        }
+
+        /// <inheritdoc cref="IDisposable"/>
+        public void Dispose()
+        {
+            stream.Dispose();
+            GC.SuppressFinalize(this);
+        }
+
+        public static IOStream CreateRingBufferStream(uint capacity)
+        {
+            return new IOStream(CIO.CreateRingBufferStream(capacity));
+        }
+
+        public static IOStream CreateFileStream(string fileName)
+        {
+            fixed (byte* fileNameBytes = ApiInterop.ToUtf8Cstr(fileName))
+            {
+                return new IOStream(CIO.CreateFileStream(fileNameBytes, CIO.OpenMode.OpenModeDefault));
+            }
+        }
+
+        public void Write(byte[] data)
+        {
+            var remainingCapacity = CIO.StreamGetRemainingWriteCapacityBytes(stream);
+            if (remainingCapacity < data.Length)
+            {
+                throw new IOException("Not enough stream capacity to write data.");
+            }
+
+            fixed (byte* dataToWrite = data)
+            {
+                var bytesWritten = CIO.StreamWrite(stream, dataToWrite, 1);
+
+                if (bytesWritten == -1)
+                {
+                    // The Stream write failed, check the error message
+                    var rawError = CIO.StreamGetLastError(stream);
+                    throw new IOException(ApiInterop.FromUtf8Cstr(rawError));
+                }
+
+                if (bytesWritten != data.Length)
+                {
+                    // Check that the number of bytes we tried to write was the number of bytes written
+                    throw new IOException("Failed to write all data to stream.");
+                }
```

Comment:
I think we should return the `bytesWritten` to the caller so they can decide what to do in this case. Right now we throw an exception, but they can't necessarily recover from it since they don't know how many bytes were written.

Reply:
Yeah, good point. I've updated some of the other functions to have a similar behaviour as this too